### ISSUE TITLE
Issue with prepareOn(), selecting(), leftJoin() and where() (combined)

### ISSUE THREAD

**Issue by Cristianmg6** [ezSQL/ezsql#197]
**Describe the bug**
I'm having problems using selecting function with prepareOn, this says the php output:
Warning: mysqli_stmt::bind_result(): Number of bind variables doesn't match number of fields in prepared statement in ****\vendor\ezsql\lib\Database\ez_mysqli.php on line 228

**To Reproduce**
$db->prepareOn()
$db->selecting() with leftJoin and where shortcut methods.

**Expected behavior**
Result with no errors, like is happening without prepareOn.

**Environment (please complete the following information):**
 - EZSQL Version : 4.0.10
 - PHP Version: 7.4.3
 - SQL Driver: mysqli

**PHP code snippet used**

```php
use ezsql\Database; 

try{
	$db = Database::initialize('mysqli', [$db_user, $db_pass, $db_name, $db_server]); //Note: vendor/driver is one of mysqli, pgsql, sqlsrv, sqlite3, or pdo.
}catch(Exception $e){
	die("Error connecting database");
}

$db->connect();
$db->prepareOn();
$db->debugOn();

$result = $db->selecting('submenu_option AS so', 'so.*, sot.*, p.name as name_page, p.id as id_page',
	leftJoin('so', 'submenu_option_text','id','id_submenu_option','sot','='),
	leftJoin('so', 'page','link_page','id','p','='),
	where(['so.id_hotel','=',1], ['sot.lang','=','es'], ['so.active','=',1]),
	orderBy('so.ord','ASC')
);

$db->debug();


```
**Generated query:**
SELECT so.*, sot.*, p.name as name_page, p.id as id_page FROM submenu_option AS so LEFT JOIN submenu_option_text AS sot ON so.id = sot.id_submenu_option LEFT JOIN page AS p ON so.link_page = p.id WHERE so.id_hotel = ? AND sot.lang = ? AND so.active = ? ORDER BY so.ord ASC

**If I don't use $db->prepareOn(), no errors reported, result data is ok and generated sql query is below:**
SELECT so.*, sot.*, p.name as name_page, p.id as id_page FROM submenu_option AS so LEFT JOIN submenu_option_text AS sot ON so.id = sot.id_submenu_option LEFT JOIN page AS p ON so.link_page = p.id WHERE so.id_hotel = '1' AND sot.lang = 'es' AND so.active = '1' ORDER BY so.ord ASC

**Tests done:**

- Same code removing both leftJoin lines and select fields, works correctly.
- Same code removing where line is working too.
- Same code with just one where condition is NOT working, same error reported.
- Also tried removing connect(), orderBy(), debug(), removing just one join, etc.

I really don't know what more can I try and I don't know what's wrong on my code :(.

**Comment by techno-express** [comment]
As a quick glance, without looking further into the methods involved signatures. The `where` should be used with the `eq()` shortcut, you really don't need `$db-connect`, it's called automatically if no connection already.  

The fact without prepareOn() no issue, leads to additional parameter not getting picked up. 
The way it works is every parameter key/value get push into array, when prepareOn, it's popped off later. 
When off everything get escaped immediately by the database vendor/driver being used.

I well look into code snipped further when I get a chance too.

**Comment by Cristianmg6** [comment]
Hi, thank you for your comments. Ok, I put $db-connect,  it just for test after see it searching on issues history.
After remove $db->connect(), this is what happens:

**With prepareOn(), same:**
Warning: mysqli_stmt::bind_result(): Number of bind variables doesn't match number of fields in prepared statement in C:\wamp64\www\fergus\backoffice\vendor\ezsql\lib\Database\ez_mysqli.php on line 228

**Without prepareOn(), not connecting:**
Fatal error: Uncaught TypeError: mysqli_real_escape_string() expects parameter 1 to be mysqli, null given in ****\vendor\ezsql\lib\Database\ez_mysqli.php:180 
Stack trace: #0 ****\vendor\ezsql\lib\Database\ez_mysqli.php(180): mysqli_real_escape_string(NULL, '1') 
#1 ****\vendor\ezsql\lib\ezQuery.php(354): ezsql\Database\ez_mysqli->escape('1')
#2 ****\vendor\ezsql\lib\ezQuery.php(454): ezsql\ezQuery->conditions('so.id_hotel', '=', 1, 'AND', NULL)
#3 ****\vendor\ezsql\lib\ezQuery.php(511): ezsql\ezQuery->processConditions('so.id_hotel', '=', 1, 'AND', NULL) 
#4 ****\vendor\ezsql\lib\ezFunctions.php(348): ezsql\ezQuery->where(Array, Array, Array) 
#5 ****\test.php(22): where(Array, Array, Array) #6 {main} thrown in ****\vendor\ezsql\lib\Database\ez_mysqli.php on line 180


**Test using the where with eq shortcut is also not working, here is the code snippet:**
```php
$result = $db->selecting('submenu_option AS so', 'so.*, sot.*, p.name as name_page, p.id as id_page',
	leftJoin('so', 'submenu_option_text','id','id_submenu_option','sot','='),
	leftJoin('so', 'page','link_page','id','p','='),
	where(eq('so.id_hotel',1, _AND), eq('sot.lang','es', _AND), eq('so.active',1)),
	orderBy('so.ord','ASC')
);
```

I'm pretty sure that I have something wrong using join's and where shortcuts together because using them in same sentence but separately, they are working fine:

**Without where conditions (result data ok):**
```php
$result = $db->selecting('submenu_option AS so', 'so.*, sot.*, p.name as name_page, p.id as id_page',
	leftJoin('so', 'submenu_option_text','id','id_submenu_option','sot','='),
	leftJoin('so', 'page','link_page','id','p','='),
	orderBy('so.ord','ASC')
);
```
_Query [1] --[SELECT so.*, sot.*, p.name as name_page, p.id as id_page FROM submenu_option AS so LEFT JOIN submenu_option_text AS sot ON so.id = sot.id_submenu_option LEFT JOIN page AS p ON so.link_page = p.id ORDER BY so.ord ASC ]_

**Without any join shortcut (result data ok):**
```php
$result = $db->selecting('submenu_option AS so', 'so.*',
	where(['so.id_hotel','=',1],['so.active','=',1]),
	orderBy('so.ord','ASC')
);
```

**Comment by techno-express** [comment]
>try{
	$db = Database::initialize('mysqli', [$db_user, $db_pass, $db_name, $db_server]); //Note: vendor/driver is one of mysqli, pgsql, sqlsrv, sqlite3, or pdo.
}catch(Exception $e){
	die("Error connecting database");
}

Why in try block?
You have a problem with database instantiation, cause you would not get any error like **Fatal error: Uncaught TypeError: mysqli_real_escape_string() expects parameter 1 to be mysqli, null**, `means no database class object found or created.`

**Comment by techno-express** [comment]
There are **Join** tests that can be used as  a **example** added by @dpDesignz 
https://github.com/ezSQL/ezsql/blob/c843017092b861e14c3f92ee2b98057d4c42f872/tests/pdo/pdo_mysqlTest.php#L288
https://github.com/ezSQL/ezsql/blob/c843017092b861e14c3f92ee2b98057d4c42f872/tests/pdo/pdo_pgsqlTest.php#L210
https://github.com/ezSQL/ezsql/blob/c843017092b861e14c3f92ee2b98057d4c42f872/tests/pdo/pdo_sqliteTest.php#L239
https://github.com/ezSQL/ezsql/blob/c843017092b861e14c3f92ee2b98057d4c42f872/tests/pdo/pdo_sqlsrvTest.php#L214

**Comment by Cristianmg6** [comment]
> > try{
> > $db = Database::initialize('mysqli', [$db_user, $db_pass, $db_name, $db_server]); //Note: vendor/driver is one of mysqli, pgsql, sqlsrv, sqlite3, or pdo.
> > }catch(Exception $e){
> > die("Error connecting database");
> > }
> 
> Why in try block?
> You have a problem with database instantiation, cause you would not get any error like **Fatal error: Uncaught TypeError: mysqli_real_escape_string() expects parameter 1 to be mysqli, null**, `means no database class object found or created.`

Phpstorm helper suggested it and I did it, but I just have test it without try block and also is reporting same error, if I put $db->connect() just below database instantiation works correctly. Is there any other way to database instantiation?

**Comment by techno-express** [comment]
There is no other way. 
If you have correct database access rights no try block necessary none of the tests use it, and connect() is called automatic by the shortcut methods, if it has proper database rights.

You do have `require 'vendor/autoload.php';` at the beginning of the file?

**Comment by techno-express** [comment]
Oh, the readme shows another way:

```
require 'vendor/autoload.php';

// Is same as:
// **** is one of mysqli, pgsql, sqlsrv, sqlite3, or Pdo.
use ezsql\Config;
use ezsql\Database\ez_****;

$setting = new Config('****', [$dsn_path_user, $password, $database, $other_settings]);
$db = new ez_****($settings);
```

**Comment by Cristianmg6** [comment]
I'm gonna checkout that Tests, is there any docs or specs about how to use it? It's just my first time using PHPUnit, I'm sorry. I tried this but it's reporting errors, if there are not any docs no problem, I'm gonna search info about PHPUnit and learn it.
```php
class MyTest extends pdo_mysqlTest{
	public function __construct(?string $name = null, array $data = [], $dataName = ''){
		parent::__construct($name, $data, $dataName);
		$this->setUp();
	}
}
$test = new MyTest();
$test->testJoins();
```


> There is no other way.
> If you have correct database access rights no try block necessary none of the tests use it, and connect() is called automatic by the shortcut methods, if it has proper database rights.
> 
> You do have `require 'vendor/autoload.php';` at the beginning of the file?

Yes, I have it.



> Oh, the readme shows another way:
> 
> ```
> require 'vendor/autoload.php';
> 
> // Is same as:
> // **** is one of mysqli, pgsql, sqlsrv, sqlite3, or Pdo.
> use ezsql\Config;
> use ezsql\Database\ez_****;
> 
> $setting = new Config('****', [$dsn_path_user, $password, $database, $other_settings]);
> $db = new ez_****($settings);
> ```

Nice, I will test it too, thank you so much.

**Comment by techno-express** [comment]
**You really need to check your setup!** 

There are no errors in the tests they all pass as constructed. They are run remote with proper database setup.
https://ci.appveyor.com/project/jv2222/ezsql
https://travis-ci.org/github/ezSQL/ezsql

**Comment by dpDesignz** [comment]
Can you try your code like this? I try to spread mine out to make it easier to read

```php
$result = $db->selecting(
    'submenu_option AS so',
    'so.*,
    sot.*,
    p.name as name_page,
    p.id as id_page',
    leftJoin(
        'so',
        'submenu_option_text',
        'id',
        'id_submenu_option',
        'sot'
    ),
    leftJoin(
        'so',
        'page',
        'link_page',
        'id',
        'p'
    ),
    where(
        eq('so.id_hotel', '1'),
        eq('sot.lang', 'es'),
        eq('so.active', '1')
    ),
    orderBy('so.ord','ASC')
);
```

The [join documentation can be found here](https://github.com/ezSQL/ezsql/wiki/Documentation#joining). It defaults to `=` so you will only need 5 arguments for it to work.

To troubleshoot any further I'd need to see your database design to understand it better.

**Comment by Cristianmg6** [comment]
Hi all, I already found the origin of the issue, it's about try to get all columns from 2 differents tables. On this example, tables 'so' and 'sot' have an 'id' field, looks like if there is 1 common field on both and you try to filter with where sentence in a prepared statement, it fails. Same is happening with 'so.*' and 'p.*'.
The example below is the same, all columns from 'sot' table except 'id' and is working correctly. 

```php
$result = $db->selecting(
	'submenu_option AS so',
	'so.*,
	 sot.id_submenu_option,
	 sot.lang,
	 sot.title,
	 sot.subtitle,
	 p.name as name_page,
         p.id as id_page',
	leftJoin(
		'so',
		'submenu_option_text',
		'id',
		'id_submenu_option',
		'sot'
	),
	leftJoin(
		'so',
		'page',
		'link_page',
		'id',
		'p'
	),
	where(
		eq('so.id_hotel', '1'),
		eq('sot.lang', 'es'),
		eq('so.active', '1')
	),
	orderBy('so.ord','ASC')
);
```

To make it easy to read, I have created two test tables with simple columns:

```sql
DROP TABLE IF EXISTS `test1`;
CREATE TABLE `test1` (
  `id` int(11) NOT NULL AUTO_INCREMENT,
  `name` varchar(255) NOT NULL,
  PRIMARY KEY (`id`)
) ENGINE=InnoDB AUTO_INCREMENT=2 DEFAULT CHARSET=latin1;

INSERT INTO `test1` (`id`, `name`) VALUES
(1,	'dad');

DROP TABLE IF EXISTS `test2`;
CREATE TABLE `test2` (
  `id` int(11) NOT NULL AUTO_INCREMENT,
  `id_test1` int(11) NOT NULL,
  `title` varchar(255) NOT NULL,
  PRIMARY KEY (`id`),
  KEY `id_test1` (`id_test1`),
  CONSTRAINT `test2_ibfk_2` FOREIGN KEY (`id_test1`) REFERENCES `test1` (`id`) ON UPDATE CASCADE
) ENGINE=InnoDB AUTO_INCREMENT=2 DEFAULT CHARSET=latin1;

INSERT INTO `test2` (`id`, `id_test1`, `title`) VALUES
(1,	1,	'son');
```

- This works:

```php
$result = $db->selecting(
	'test1 AS t1',
	't1.*,
         t2.title',
	leftJoin(
		't1',
		'test2',
		'id',
		'id_test1',
		't2'
	),
	where(
		eq('t1.id', '1')
	)
);

```

- And this not:

```php
$result = $db->selecting(
	'test1 AS t1',
	't1.*,
         t2.*',
	leftJoin(
		't1',
		'test2',
		'id',
		'id_test1',
		't2'
	),
	where(
		eq('t1.id', '1')
	)
);

```

As you can see, the only one difference is in 4th line, 't2.title' vs 't2.*'.


> **You really need to check your setup!**
> 
> There are no errors in the tests they all pass as constructed. They are run remote with proper database setup.
> https://ci.appveyor.com/project/jv2222/ezsql
> https://travis-ci.org/github/ezSQL/ezsql

I know there are no erros, I just said that I don't know how to run it because is my first time using PHPUnit, but as I also said, I'm gonna check it to learn and know how to do it, I'm sorry, thanks.





> Can you try your code like this? I try to spread mine out to make it easier to read

Yes, done.

> The [join documentation can be found here](https://github.com/ezSQL/ezsql/wiki/Documentation#joining). It defaults to `=` so you will only need 5 arguments for it to work.

Done too, thank you.

> To troubleshoot any further I'd need to see your database design to understand it better.

I had put the database schema with two test tables where I'm having same issue.

I'm sorry if I'm doing something wrong, last thing I want is waste your time, and thank you so much for this great library.

**Comment by dpDesignz** [comment]
@Cristianmg6 Great. Glad to hear that you got it sorted.

Don't worry about not understanding PHPUnit testing. I had never used it until I started using this project and it literally took me a couple of months to wrap my head around it. I wrote [this in the troubleshooting of the Wiki](https://github.com/ezSQL/ezsql/wiki/Troubleshooting#how-to-run-a-phpunit-test) after I finally worked it out. See how you get on with that. 😄 

On that, just a recommendation for your database design, here's a couple of great articles you can read around design concepts ([10 Best Database Design Practices](https://medium.com/quick-code/10-best-database-design-practices-1f10f3441730), [SQL naming conventions](https://www.red-gate.com/simple-talk/blogs/sql-naming-conventions/)). I keep away from camel casing my names as I've had issues between some databases not allowing them.
My biggest suggestion for you though if you're not wanting to add a prefix to your column names, at least add a prefix to your `id` column so you don't have issues like this popping up. Personally, I prefix all my columns to make it easier for myself. So for `submenu_option.id`, I would call that `submenu_option.smo_id`. And `submenu_option_text.id`, I would call that `submenu_option_text.smot_id`. Just a suggestion for you to try 😄 

We're all here to learn. As long as we can see you're attempting to work it out on your own, most people will be more than happy to help. Best of luck!

**Comment by Cristianmg6** [comment]
@dpDesignz Thank you so much for your comments, I appreciate it.
That docs about PHPUnit testing is just what I were looking for, I'm gonna test it and also read offical docs tu learn it, thanks.

About the database design, I also appreciate your comments and links about good practices, I will keep it in mind to next projects. 

The problem here is that the reason I'm using is to change the one I had before (just and older version of ezsql from some years ago). That means that I already have the database created and can not change it now.
So I just want to ask if is there any other way to use this version and get all params from tables using prepared statements, instead of select all columns one by one.

**Comment by dpDesignz** [comment]
> So I just want to ask if is there any other way to use this version and get all params from tables using prepared statements, instead of select all columns one by one.

There's a good [answer on SO](https://stackoverflow.com/a/23507737/1248664) about this. While it is tempting to using the select all query, it is safer, faster, and more secure to select the columns that you actually need. Yes, it takes a couple more minutes typing to write out your query, but it's a lot easier than spending hours trying to stop a bug or leak down the track because you ended up with another duplicate column name or something happened to your table.

Since your question is now sorted I'd suggest you mark this as closed and if you want to continue this discussion I suggest you post on Stack Overflow or the likes and you'll be able to get specific SQL help as these are SQL questions, not ezSQL specific 😄

**Comment by Cristianmg6** [comment]
@dpDesignz Thank you again, Damien, I have used select * in a lot of queries but never with prepared statements and join's, so I didn't know about this issue when having same name columns, it's the reason I thought that the problem was on the library instead of in my database design.
Thank you both for your help, suggestions and for the library, and sorry for reporting a non-existent bug.
I proceed to close.
